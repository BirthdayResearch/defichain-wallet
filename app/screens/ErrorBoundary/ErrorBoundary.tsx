import React, { PropsWithChildren } from 'react'
import { Text, View } from '../../components'
import { AppIcon } from '../../components/icons/AppIcon'
import { tailwind } from '../../tailwind'
import { translate } from '../../translations'

function ErrorScreen (): JSX.Element {
  return (
    <View style={tailwind('flex-1 items-center justify-center p-4')}>
      <AppIcon />
      <Text style={tailwind('text-xl font-bold mt-3')}>
        {translate('screens/ErrorBoundary', 'Something went wrong')}
      </Text>
      <Text style={tailwind('text-sm text-gray-400 font-bold mt-3 text-center')}>
        {translate('screens/ErrorBoundary', 'The app ran into a problem and could not recover. Please restart your application.')}
      </Text>
    </View>
  )
}

interface ErrorBoundaryState {
  hasError: boolean
}

export class ErrorBoundary extends React.Component<PropsWithChildren<{}>, ErrorBoundaryState> {
  constructor (props: PropsWithChildren<{}>) {
    super(props)
    this.state = { hasError: false }
  }

  static getDerivedStateFromError (): ErrorBoundaryState {
    return { hasError: true }
  }

  render (): JSX.Element {
    if (this.state.hasError) {
      return <ErrorScreen />
    }

    return (
      <>
        {this.props.children}
      </>
    )
  }
}
