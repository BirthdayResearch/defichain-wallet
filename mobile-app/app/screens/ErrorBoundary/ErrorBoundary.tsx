import { Component, ReactElement } from 'react'
import { Text, View } from '../../components'
import { AppIcon } from '@components/icons/AppIcon'
import { tailwind } from '@tailwind'
import { translate } from '@translations'

interface Props {
  children: ReactElement
}

interface State {
  hasError: boolean
}

export function ErrorDisplayComponent (): JSX.Element {
  return (
    <View style={tailwind('flex-1 items-center justify-center p-4')}>
      <AppIcon />

      <Text style={tailwind('text-xl font-bold mt-3')}>
        {translate('screens/ErrorBoundary', 'Something went wrong')}
      </Text>

      <Text style={tailwind('text-sm text-gray-400 font-bold mt-3 text-center')}>
        {translate('screens/ErrorBoundary', 'The app ran into a problem and could not recover. Please restart your application.')}
      </Text>
    </View>
  )
}

class ErrorBoundary extends Component<Props, State> {
  constructor (props: Props) {
    super(props)
    this.state = { hasError: false }
  }

  static getDerivedStateFromError (): State {
    return { hasError: true }
  }

  render (): JSX.Element {
    return this.state.hasError
      ? <ErrorDisplayComponent />
      : this.props.children
  }
}

export default ErrorBoundary
