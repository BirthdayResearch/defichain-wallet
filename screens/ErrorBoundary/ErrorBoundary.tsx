import React, { ReactElement } from 'react'
import { Text, View } from '../../components/Themed'
import tailwind from 'tailwind-rn'
import { translate } from '../../translations'

interface Props {
  children: ReactElement
}

interface State {
  hasError: boolean
}

export function ErrorDisplayComponent (): JSX.Element {
  return (
    <View style={tailwind('flex-1 items-center justify-center')}>
      <Text style={tailwind('text-xl font-bold')}>
        {translate('screens/ErrorBoundary', 'Something went wrong')}
      </Text>
    </View>
  )
}

class ErrorBoundary extends React.Component<Props, State> {
  constructor (props: Props) {
    super(props)
    this.state = { hasError: false }
  }

  static getDerivedStateFromError (): State {
    return { hasError: true }
  }

  render (): JSX.Element {
    return this.state.hasError
      ? <ErrorDisplayComponent />
      : this.props.children
  }
}

export default ErrorBoundary
