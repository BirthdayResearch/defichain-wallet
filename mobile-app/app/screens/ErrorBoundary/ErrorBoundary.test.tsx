import { render } from '@testing-library/react-native'
import { Text, View } from 'react-native'
import ErrorBoundary from './ErrorBoundary'

describe('ErrorBoundary', () => {
  let consoleErrorSpy: jest.SpyInstance
  let errorMock: Error

  beforeAll(() => {
    consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => {
    })
    errorMock = new Error('This is a test error!')
  })

  afterAll(() => {
    consoleErrorSpy.mockRestore()
  })

  describe('when there is no error', () => {
    it('<ErrorBoundary /> should render children components', () => {
      const tree = render(
        <ErrorBoundary>
          <Text>
            'Child Component'
          </Text>
        </ErrorBoundary>).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('when there is an error', () => {
    it('<ErrorBoundary /> should render when there is an error', () => {
      const wrapper = render(
        <ErrorBoundary>
          <View>
            {errorMock}
          </View>
        </ErrorBoundary>
      )
      expect(wrapper).toMatchSnapshot()
    })
  })
})
