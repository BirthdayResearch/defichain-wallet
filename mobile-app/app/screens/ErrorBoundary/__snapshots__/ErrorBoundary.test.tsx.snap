// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary when there is an error <ErrorBoundary /> should render when there is an error 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexBasis": "0%",
      "flexGrow": 1,
      "flexShrink": 1,
      "justifyContent": "center",
      "paddingBottom": 16,
      "paddingLeft": 16,
      "paddingRight": 16,
      "paddingTop": 16,
    }
  }
>
  <RNSVGSvgView
    align="xMidYMid"
    bbHeight="64"
    bbWidth="64"
    focusable={false}
    height="64"
    meetOrSlice={0}
    minX={0}
    minY={0}
    style={
      Array [
        Object {
          "backgroundColor": "transparent",
          "borderWidth": 0,
        },
        Object {
          "flex": 0,
          "height": 64,
          "width": 64,
        },
      ]
    }
    vbHeight={1024}
    vbWidth={1024}
    width="64"
  >
    <RNSVGGroup>
      <RNSVGGroup
        fill={null}
        fillRule={0}
        name="icon-1024"
        propList={
          Array [
            "fill",
            "fillRule",
            "stroke",
            "strokeWidth",
          ]
        }
        stroke={null}
        strokeWidth="1"
      >
        <RNSVGRect
          fill={4294698742}
          height="896"
          name="Rectangle"
          propList={
            Array [
              "fill",
            ]
          }
          rx="448"
          width="896"
          x="64"
          y="64"
        />
        <RNSVGPath
          d="M573,717 L573,309 C660.2327,335.241169 724,416.762111 724,513 C724,609.237889 660.2327,690.774035 573,717 M511.924146,239 L511.924146,469.0935 L477.258683,434.437667 L472.828786,347.942167 L508.980995,239.151667 C485.799889,239.394333 463.407669,242.700667 441.971214,248.388167 L424.479189,300.971 L374.90075,276.112833 C355.284801,287.533333 337.140428,301.1985 321.013782,316.987 L412.843123,362.9875 L415.528369,415.631 L362.885413,412.931333 L316.85696,321.142667 C301.079244,337.264833 287.39511,355.389 275.971437,375.014667 L300.851681,424.5945 L248.223896,442.0665 C242.549986,463.497 239.242734,485.883 239,509.057667 L347.835899,472.9155 L434.355432,477.329 L469.020895,512 L434.355432,546.671 L347.835899,551.0845 L239,514.942333 C239.242734,538.117 242.549986,560.503 248.223896,581.9335 L300.851681,599.4055 L275.971437,648.985333 C287.39511,668.611 301.079244,686.735167 316.85696,702.857333 L362.885413,611.068667 L415.528369,608.369 L412.843123,660.997333 L321.013782,707.013 C337.140428,722.786333 355.284801,736.4515 374.90075,747.887167 L424.479189,723.029 L441.971214,775.611833 C463.407669,781.284167 485.799889,784.605667 508.980995,784.848333 L472.828786,676.057833 L477.258683,589.562333 L511.924146,554.9065 L511.924146,785 C662.737872,785 785,662.771833 785,512 C785,361.228167 662.737872,239 511.924146,239"
          fill={4294901935}
          name="Fill-1"
          propList={
            Array [
              "fill",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGGroup>
  </RNSVGSvgView>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "RegularFont",
          "fontSize": 16,
          "fontWeight": "400",
          "lineHeight": 24,
        },
        Object {
          "fontFamily": "BoldFont",
          "fontSize": 20,
          "fontWeight": "700",
          "lineHeight": 28,
          "marginTop": 12,
        },
      ]
    }
  >
    Something went wrong
  </Text>
  <Text
    style={
      Array [
        Object {
          "fontFamily": "RegularFont",
          "fontSize": 16,
          "fontWeight": "400",
          "lineHeight": 24,
        },
        Object {
          "color": "rgba(163, 163, 163, 1)",
          "fontFamily": "BoldFont",
          "fontSize": 14,
          "fontWeight": "700",
          "lineHeight": 20,
          "marginTop": 12,
          "textAlign": "center",
        },
      ]
    }
  >
    The app ran into a problem and could not recover. Please restart your application.
  </Text>
</View>
`;

exports[`ErrorBoundary when there is no error <ErrorBoundary /> should render children components 1`] = `
<Text>
  'Child Component'
</Text>
`;
